Fix byte unit: 1024 B should print as 1 KiB. (copy pr #19040)

inkydragon opened the pull request:
copy pr #19040 which cannot be merge.
and add test for `factor = 1000` too.

fix #19039

inkydragon committed:
1024 B should show as 1 KiB

copy pr #19040

inkydragon committed:
Add test set for mem_units

inkydragon committed:
Add test set for cnt_units

inkydragon committed:
fix typo

inkydragon committed:
Merge branch 'master' into fix-byte-unit

inkydragon committed:
fix Trailing whitespace

- check use r" $"

JeffBezanson commented on base/util.jl:
```diff
@@ -68,13 +68,12 @@ gc_bytes() = ccall(:jl_gc_total_bytes, Int64, ())
-    if value == 0 || value == 1
-        return (value, 1)
+    unit = 1
+    while value >= factor && unit < numunits
```
This should be made type stable, e.g. by putting `value /= one(factor)` before the loop.

stevengj replied:
When pretty-printing, it would be nice to omit the `.0` for exact numbers of bytes.  It seems weird to write `"998.0 bytes"`, for example.

Of course, that could be done in the printing routine rather than here, but I'm not sure type-stability is a big issue here.

We could always do it at the end by changing the return value to `return value == floor(value) ? Int(value) : value, unit`, which would also print e.g. 1024 bytes as "1 KiB" rather than "1.0 KiB".

inkydragon committed:
make code type stable

inkydragon committed:
Merge branch 'master' into fix-byte-unit

inkydragon committed:
fix test error

del `const`: fix error "syntax: unsupported  `const` declaration on local variable"

oscardssmith commented:
Closing as `@time` now does this right.
